rfqservice/rfq.go

```diff
@@ -0,0 +1,234 @@
+package rfqservice
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/lightninglabs/taproot-assets/fn"
+	msg "github.com/lightninglabs/taproot-assets/rfqmessages"
+)
+
+const (
+	// DefaultTimeout is the default timeout used for context operations.
+	DefaultTimeout = 30 * time.Second
+)
+
+// ManagerCfg is a struct that holds the configuration parameters for the RFQ
+// manager.
+type ManagerCfg struct {
+	// PeerMessagePorter is the peer message porter. This component
+	// provides the RFQ manager with the ability to send and receive raw
+	// peer messages.
+	PeerMessagePorter PeerMessagePorter
+}
+
+// Manager is a struct that handles the request for quote (RFQ) system.
+type Manager struct {
+	startOnce sync.Once
+	stopOnce  sync.Once
+
+	// cfg holds the configuration parameters for the RFQ manager.
+	cfg ManagerCfg
+
+	// rfqStreamHandle is the RFQ stream handler. This subsystem handles
+	// incoming and outgoing peer RFQ stream messages.
+	rfqStreamHandle *StreamHandler
+
+	// quoteNegotiator is the RFQ quote negotiator. This subsystem
+	// determines whether a quote is accepted or rejected.
+	quoteNegotiator *QuoteNegotiator
+
+	// ContextGuard provides a wait group and main quit channel that can be
+	// used to create guarded contexts.
+	*fn.ContextGuard
+}
+
+// NewManager creates a new RFQ manager.
+func NewManager(cfg ManagerCfg) (Manager, error) {
+	return Manager{
+		cfg: cfg,
+		ContextGuard: &fn.ContextGuard{
+			DefaultTimeout: DefaultTimeout,
+			Quit:           make(chan struct{}),
+		},
+	}, nil
+}
+
+// Start attempts to start a new RFQ manager.
+func (m *Manager) Start() error {
+	var startErr error
+	m.startOnce.Do(func() {
+		ctx, cancel := m.WithCtxQuitNoTimeout()
+
+		log.Info("Initializing RFQ subsystems")
+		err := m.startSubsystems(ctx)
+		if err != nil {
+			startErr = err
+			return
+		}
+
+		// Start the manager's main event loop in a separate goroutine.
+		m.Wg.Add(1)
+		go func() {
+			defer func() {
+				m.Wg.Done()
+
+				// Attempt to stop all subsystems if the main
+				// event loop exits.
+				err = m.stopSubsystems()
+				if err != nil {
+					log.Errorf("Error stopping RFQ "+
+						"subsystems: %v", err)
+				}
+
+				// The context can now be cancelled as all
+				// dependant components have been stopped.
+				defer cancel()
+			}()
+
+			log.Info("Starting RFQ manager main event loop")
+			err = m.mainEventLoop()
+			if err != nil {
+				startErr = err
```

Comment:
This `startErr` will never be evaluated as we're in a goroutine. I think we should instead just do `go m.mainEventLoop()` and do all the `defer` and error handling there.